---
title: "Examples utilizing the DNA methylation and RNA-seq data provided."
author: "Charles Imbusch"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## TWGBS data

The DNA methylation data have been stored as rds files and are bsseq objects. They contain the raw counts for each CpG position in the mm10 reference genome as well the preprocessed smoothed values for each sample. In addition there is a second file containing DNA methylation profiles per tissue, e.g. the replicates from the same tissue have been collapsed. Also here the smoothed values have been precomputed and are directly available.

### Example utilization

We first load necessary libraries and data:

```{r include = FALSE}
library(bsseq)
```

```{r echo = TRUE}
BS.obj.ex.fit <- readRDS(file = "/icgc/dkfzlsdf/analysis/G200/imbusch/treg/treg_experimentHub/treg_twgbs_per_group.rds")
```

For example to visually inspect the DNA methylation state of the FoxP3 gene in the CNS2 region we first specify a single region:

```{r echo = TRUE}
regions <- GRanges(
  seqnames = c("X"),
  ranges = IRanges(start = c(7579676),
                   end = c(7595243)
  )
)
```

We can then directly use the plotRegion function from the bsseq package for visualization:

```{r echo = TRUE}
plotRegion(BS.obj.ex.fit, region=regions[1,], extend = 2000)
```
